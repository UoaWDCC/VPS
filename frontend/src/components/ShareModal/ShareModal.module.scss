.dialogTitle {
  color: #fafafa;
  background-color: #2a555e;
}
.dialogContent {
  color: #fafafa;
  background-color: #0080a7;
  display: flex;
  flex-direction: column;
  align-items: center;

  > .dialogItem {
    margin-block: 5px;
  }
}

.textField {
  background-color: #fafafa;
  border-radius: 10px;
}
