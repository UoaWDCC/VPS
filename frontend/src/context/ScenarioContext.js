import React from "react";

const ScenarioContext = React.createContext();

export default ScenarioContext;
