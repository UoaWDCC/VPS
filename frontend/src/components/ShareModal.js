import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import Typography from "@material-ui/core/Typography";
import React, { useContext, useState } from "react";
import { DialogContent, DialogTitle, TextField } from "@material-ui/core";
import ScenarioContext from "../context/ScenarioContext";
import styles from "../styling/ShareModal.module.scss";

export default function ShareModal({ isOpen, handleClose }) {
  const { currentScenario } = useContext(ScenarioContext);
  const [copySuccess, setCopySuccess] = useState(false);
  const url = `${window.location.origin}/play/${currentScenario._id}`;

  function onClose() {
    handleClose();
    setCopySuccess(false);
  }

  return (
    <div>
      <Dialog
        className={styles.dialog}
        onClose={onClose}
        open={isOpen}
        maxWidth="sm"
      >
        <DialogTitle className={styles.dialogTitle}>Shareable Link</DialogTitle>
        <DialogContent className={styles.dialogContent}>
          <Typography className={styles.dialogItem}>
            Give this link to others so they can play this scenario
          </Typography>
          <TextField
            className={`${styles.textField} ${styles.dialogItem}`}
            defaultValue={url}
            InputProps={{
              readOnly: true,
            }}
            variant="outlined"
            fullWidth
          />
          <Button
            className={`btn contained white ${styles.dialogItem}`}
            autoFocus
            onClick={() => {
              navigator.clipboard.writeText(url);
              setCopySuccess(true);
            }}
            color="primary"
          >
            {copySuccess ? "Copied!" : "Copy Link"}
          </Button>
        </DialogContent>
      </Dialog>
    </div>
  );
}
