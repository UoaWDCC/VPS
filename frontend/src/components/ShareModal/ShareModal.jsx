import Button from "@material-ui/core/Button";
import { useContext, useState } from "react";
import ScenarioContext from "../../context/ScenarioContext";
import styles from "./ShareModal.module.scss";

/**
 * Component used to a display a share model on the screen, conisting of a copiable link and a button.
 *
 * @component
 * @example
 * const isOpen = false
 * function handleClose() {
 *   console.log("Closed.")
 * }
 * return (
 *   <ShareModal isOpen={isOpen} handleClose={handleClose} />
 * )
 */
export default function ShareModal({ isOpen, handleClose }) {
  const { currentScenario } = useContext(ScenarioContext);
  const [copySuccess, setCopySuccess] = useState(false);
  const url = `${window.location.origin}/play/${currentScenario._id}`;

  /** Function which executes when the modal is closed. */
  function onClose() {
    handleClose();
    setCopySuccess(false);
  }

  return (
    <div>
      <dialog
        id="share_modal"
        className={`modal ${isOpen ? "modal-open" : ""}`}
      >
        <form
          method="dialog"
          className="modal-box relative flex flex-col items-center justify-center"
          style={{
            backgroundColor: "white",
            color: "black",
            borderRadius: "10px",
          }}
        >
          <button
            className="btn btn-sm btn-circle btn-ghost absolute right-2 top-2"
            onClick={onClose}
          >
            âœ•
          </button>
          <h3 className="font-bold text-lg text-center mb-4 text-black">
            Shareable Link
          </h3>
          <div className="w-full">
            <p className="text-center text-black mb-4">
              Give this link to others so they can play this scenario
            </p>
            <input
              type="text"
              value={url}
              readOnly
              onFocus={(e) => e.target.select()}
              className="input input-bordered w-full mb-4 text-center"
              style={{
                backgroundColor: "white",
                color: "black",
                borderRadius: "5px",
              }}
            />
            <div className="flex justify-center">
              <Button
                className={`btn contained white ${styles.dialogItem}`}
                autoFocus
                onClick={() => {
                  navigator.clipboard.writeText(url);
                  setCopySuccess(true);
                }}
                color="primary"
              >
                {copySuccess ? "Copied!" : "Copy Link"}
              </Button>
            </div>
          </div>
          <div className="modal-action"></div>
        </form>
      </dialog>
    </div>
  );
}
