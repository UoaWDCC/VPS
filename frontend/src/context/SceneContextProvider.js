import React, { useContext, useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import ScenarioContext from "./ScenarioContext";
import SceneContext from "./SceneContext";

export default function SceneContextProvider({ children }) {
  const { currentScenario } = useContext(ScenarioContext);
  const [scenes, setScenes] = useState([]);
  const [currentScene, setCurrentScene] = useLocalStorage("currentScene", null);
  let getScenes = null;
  if (currentScenario) {
    getScenes = useGet(`api/scenario/${currentScenario._id}/scene`, setScenes);
  }

  return (
    <SceneContext.Provider
      value={{
        scenes,
        setScenes,
        reFetch: getScenes?.reFetch,
        currentScene,
        setCurrentScene,
      }}
    >
      {children}
    </SceneContext.Provider>
  );
}
