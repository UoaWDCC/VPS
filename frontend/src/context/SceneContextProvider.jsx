import { useContext, useEffect, useRef, useState } from "react";
import { useGet } from "../hooks/crudHooks";
import useLocalStorage from "../hooks/useLocalStorage";
import AuthenticationContext from "./AuthenticationContext";
import ScenarioContext from "./ScenarioContext";
import SceneContext from "./SceneContext";
import { useParams } from "react-router-dom";
import { api } from "../util/api";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import LoadingPage from "../features/status/LoadingPage";
import GenericErrorPage from "../features/status/GenericErrorPage";
import toast from "react-hot-toast";
import useVisualScene from "../features/authoring/stores/visual";
import { getScene } from "../features/authoring/scene/scene";

async function getAllScenes(user, id) {
  const res = await api.get(user, `api/scenario/${id}/scene/all`)
  return res.data.filter(Boolean);
}

function updateScenes(user, id, sceneIds) {
  api.put(
    user,
    `/api/scenario/${id}/scene/reorder`,
    { sceneIds });
}

function deleteScene(user, scenarioId, sceneId) {
  api.delete(user, `/api/scenario/${scenarioId}/scene/${sceneId}`);
};

async function saveScene(user, scenarioId, scene) {
  const components = Object.values(scene.components);
  await api.put(user, `/api/scenario/${scenarioId}/scene/${scene._id}`, { ...scene, components });
}

/**
 * This is a Context Provider made with the React Context API
 * SceneContextProvider allows access to scene info and the refetch function
 */
export default function SceneContextProvider({ children }) {
  const { setCurrentScenario } = useContext(ScenarioContext);
  const { user } = useContext(AuthenticationContext);
  const { scenarioId } = useParams();

  const queryClient = useQueryClient();

  // TODO: this is disgusting
  const scenarioQuery = useQuery({
    queryKey: ["scenario", scenarioId],
    queryFn: () => api.get(user, `api/scenario/${scenarioId}`).then(r => r.data),
    enabled: !!scenarioId,
    onSuccess: (data) => setCurrentScenario(data),
  });

  const scenesQuery = useQuery({
    queryKey: ["scenes", scenarioId],
    queryFn: () => getAllScenes(user, scenarioId),
    enabled: !!scenarioId
  });

  const reorderMutation = useMutation({
    mutationFn: (ids) => updateScenes(user, scenarioId, ids),
    onMutate: async (ids) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], (prev = []) => {
        return ids.map(id => prev.find(s => s._id === id));
      });
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      // queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  const deleteMutation = useMutation({
    mutationFn: (id) => deleteScene(user, scenarioId, id),
    onMutate: async (id) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], prev => prev ? prev.filter(s => s._id !== id) : []);
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  const saveSceneMutation = useMutation({
    mutationFn: (scene) => saveScene(user, scenarioId, scene),
    onMutate: async (scene) => {
      await queryClient.cancelQueries(["scenes", scenarioId]);
      queryClient.setQueryData(["scenes", scenarioId], (prev = []) => {
        const index = prev.findIndex(s => s._id === scene._id);
        return index === -1 ? prev : prev.toSpliced(index, 1, scene);
      })
    },
    onError: (_err, _updates, context) => {
      toast.error("Something went wrong updating the scenes, your last changes weren't saved");
    },
    onSettled: () => {
      queryClient.invalidateQueries(["scenes", scenarioId]);
    }
  });

  if (scenesQuery.isLoading) {
    return <LoadingPage text="Getting scenes..." />;
  }

  if (scenesQuery.isError) {
    console.error(scenesQuery.error);
    return <GenericErrorPage />;
  }

  return (
    <SceneContext.Provider
      value={{
        scenes: scenesQuery.data,
        reorderScenes: reorderMutation.mutate,
        saveScene: saveSceneMutation.mutate,
        deleteScene: deleteMutation.mutate,
        reFetch: scenesQuery.refetch,
      }}
    >
      {children}
    </SceneContext.Provider>
  );
}
