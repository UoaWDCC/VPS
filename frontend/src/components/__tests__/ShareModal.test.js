import React from "react";
import { BrowserRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import ShareModal from "../ShareModal";
import ScenarioContext from "../../context/ScenarioContext";
import SceneContextProvider from "../../context/SceneContextProvider";

test("ShareModal component snapshot test", () => {
  const context = {
    currentScenario: { _id: "test" },
  };

  const { baseElement } = render(
    <BrowserRouter>
      <ScenarioContext.Provider value={context}>
        <SceneContextProvider>
          <ShareModal isOpen />
        </SceneContextProvider>
      </ScenarioContext.Provider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});
