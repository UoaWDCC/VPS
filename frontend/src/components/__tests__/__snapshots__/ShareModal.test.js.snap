// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareModal component snapshot test 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div />
  </div>
  <div
    class="MuiDialog-root dialog"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root dialogTitle"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Shareable Link
          </h2>
        </div>
        <div
          class="MuiDialogContent-root dialogContent"
        >
          <p
            class="MuiTypography-root dialogItem MuiTypography-body1"
          >
            Give this link to others so they can play this scenario
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root textField dialogItem MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                readonly=""
                type="text"
                value="http://localhost/play/test"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-2"
                  style="width: 0.01px;"
                >
                  <span>
                    â€‹
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text btn contained white dialogItem MuiButton-textPrimary Mui-focusVisible Mui-focusVisible"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Copy Link
            </span>
            <span
              class="MuiTouchRipple-root"
            >
              <span
                class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
